### ISSUE TITLE
Problems with text encoding of http response (European characters)

### ISSUE THREAD

**Issue by runesig** [boostorg/beast#1083]
I am trying out the Beast library but I have a couple of issues. I am struggeling to get the response from the server in correct format under Windows. I have tried the examples that follows with the library with a test server, where:

http::request<http::string_body> req_;
http::response<http::string_body> res_;

I am using JSON and a stringstream like this:

std::stringstream message_string;
write_json(message_string, message);

and:

req_.body() = message_string.str();

If I try to post a name like "René Andersson" the request is correct and the server gets the characters in the correct format. 

BUT, the response (which is the exact same text) is not correct. I pick out the body from the response like this:

std::stringstream message_string(response.body());

The name gets changed to René = Renï¿½ 

Can you please help tell me what I am doing wrong?

**Comment by carolinebeltran** [comment]
A quick suggestion, I would suggest setting your charset to utf-8.  For example:

`res.set(http::field::content_type, "application/json; charset=utf-8");`

I do this and can display latin, Chinese, Russian, etc. characters just fine for my web application, reading from and writing to text or JSON files.  Although UTF-8 has no problems with Windows, trying to print multi-byte UTF-8 characters (such as Russian or Chinese for example) to a Windows command prompt will result in output similar to what you printed in your post.

**Comment by carolinebeltran** [comment]
Of course, if you are going to display as HTML, you would use:

`res.set(http::field::content_type, "html/text; charset=utf-8");`

**Comment by runesig** [comment]
Thanks, for your suggestion. I had not tried to set it on the response until now, but I just did and unfortunately the result is the same. I am not printing the value to a console but using it for a search.  And, of course the search is failing because of this. Also, the debugger shows me the same result. I am using Visual Studio, changing character set from unicode to multi byte does not do any difference either.

**Comment by vinniefalco** [comment]
>Can you please help tell me what I am doing wrong?

Can you please post a complete program?

**Comment by runesig** [comment]
[BeastWindowsTestProject.zip](https://github.com/boostorg/beast/files/1866598/BeastWindowsTestProject.zip)

Yes, I have attached a very simple Win test project :) Thanks!

**Comment by runesig** [comment]
Btw, this same project is what I used to test the async issue.

**Comment by vinniefalco** [comment]
The program looks okay, although it could have been trimmed to just the parts needed to reproduce the defect. Is the server sending back an incorrect result?

**Comment by runesig** [comment]
Yes, I am sorry, I just sent you my test as was. The server is responding correctly according to Postman. The string in the response object is the problem.

**Comment by vinniefalco** [comment]
Beast doesn't perform any translation of the message body, so I don't see how it is possible that there is a bug in the library causing this.

**Comment by runesig** [comment]
So that means that it’s either me who have done something wrong, or an issue with Asio?

**Comment by vinniefalco** [comment]
There's no way that Asio is causing this, it has to be either in your program or the server.

**Comment by runesig** [comment]
Yes you are right. I have done some more research. I tried another library based on libcurl (Cpr) and got the exact same issue. I also made a C# program where I used HttpClient and PostAsync where everything was correct. So it is not the server (I already knew that), it is some kind of a c++ windows character set/ encoding issue.

**Comment by vinniefalco** [comment]
Maybe you need to `imbue` the `std::stringstream` with the "C" locale?

**Comment by runesig** [comment]
Thanks for your suggestion, but I could not get it to work. But since this is not directly related Beast I close this issue and I'll ask the question on Stackoverflow too. Thanks!

**Comment by runesig** [comment]
Sometimes, its so obvious when you finally figure out what the problem really was... I have been looking at the wrong place all the time. But thats how things work from time to time. The string with the European characters I was sending was in iso-8859-1 format, not utf-8. I had to convert it first, then post it, and then convert the response back. I used the boost library like this 

std::string output = boost::locale::conv::to_utf<char>(someString, loc);

By setting "content_type" to "charset=utf-8" does of course not magically convert the string, it just tells the server which format the string you are sending is in. 

In C# we can set :
var content = new StringContent(jsonObject.ToString(), Encoding.UTF8, "application/json");
The Encoding.UTF8 parameter does convert the text and I guess that is what fooled me when I compared the results.

Thanks @carolinebeltran and @vinniefalco for looking into this. Appreciate it!
